File: src/main/java/com/lethimcook/Converter.java
package com.lethimcook;

/**
 * Core unit conversion functionality.
 * Provides methods to convert values between different units of the same type.
 */
public final class Converter {

    private Converter() {
    }

    /**
     * Convert a value from one unit to another.
     *
     * @param value    the numeric value to convert
     * @param fromUnit the source unit
     * @param toUnit   the target unit
     * @return the converted value
     * @throws IllegalArgumentException if units are incompatible or unknown
     */
    public static double convert(double value, String fromUnit, String toUnit) {
        String normalizedFrom = Units.normalizeUnit(fromUnit);
        String normalizedTo = Units.normalizeUnit(toUnit);

        // Get unit types
        UnitType fromType = Units.getUnitType(normalizedFrom);
        UnitType toType = Units.getUnitType(normalizedTo);

        // Check compatibility
        if (fromType != toType) {
            throw new IllegalArgumentException(
                    "Cannot convert between " + fromType + " and " + toType
            );
        }

        // Handle temperature separately (non-linear conversion)
        if (fromType == UnitType.TEMPERATURE) {
            return convertTemperature(value, normalizedFrom, normalizedTo);
        }

        // Handle count (no conversion needed)
        if (fromType == UnitType.COUNT) {
            return value;
        }

        // Convert: from_unit -> base_unit -> to_unit
        double baseValue = value * Units.CONVERSIONS.get(normalizedFrom);
        return baseValue / Units.CONVERSIONS.get(normalizedTo);
    }

    /**
     * Convert temperature between different scales.
     * Uses Celsius as the intermediate scale for all conversions.
     *
     * @param value    the temperature value to convert
     * @param fromUnit the source temperature unit (normalized)
     * @param toUnit   the target temperature unit (normalized)
     * @return the converted temperature value
     */
    private static double convertTemperature(double value, String fromUnit, String toUnit) {
        // First convert to Celsius
        double celsius = switch (fromUnit) {
            case "celsius", "c" -> value;
            case "fahrenheit", "f" -> (value - 32) * 5 / 9;
            case "kelvin", "k" -> value - 273.15;
            default -> throw new IllegalArgumentException("Unknown temperature unit: " + fromUnit);
        };

        // Then convert from Celsius to target
        return switch (toUnit) {
            case "celsius", "c" -> celsius;
            case "fahrenheit", "f" -> celsius * 9 / 5 + 32;
            case "kelvin", "k" -> celsius + 273.15;
            default -> throw new IllegalArgumentException("Unknown temperature unit: " + toUnit);
        };
    }
}

File: src/converter.cpp
#include "lethimcook/converter.hpp"
#include "lethimcook/units.hpp"

#include <stdexcept>

namespace lethimcook {

// ---------------------------------------------------------------------------
// Internal: temperature conversion via Celsius as intermediate.
// ---------------------------------------------------------------------------
static double convert_temperature(double value,
                                  const std::string& from_unit,
                                  const std::string& to_unit) {
    // Step 1: convert to Celsius
    double celsius = 0.0;
    if (from_unit == "celsius" || from_unit == "c") {
        celsius = value;
    } else if (from_unit == "fahrenheit" || from_unit == "f") {
        celsius = (value - 32.0) * 5.0 / 9.0;
    } else if (from_unit == "kelvin" || from_unit == "k") {
        celsius = value - 273.15;
    } else {
        throw std::invalid_argument("Unknown temperature unit: " + from_unit);
    }

    // Step 2: convert from Celsius to target
    if (to_unit == "celsius" || to_unit == "c") {
        return celsius;
    } else if (to_unit == "fahrenheit" || to_unit == "f") {
        return celsius * 9.0 / 5.0 + 32.0;
    } else if (to_unit == "kelvin" || to_unit == "k") {
        return celsius + 273.15;
    } else {
        throw std::invalid_argument("Unknown temperature unit: " + to_unit);
    }
}

// ---------------------------------------------------------------------------
// convert
// ---------------------------------------------------------------------------
double convert(double value,
               const std::string& from_unit,
               const std::string& to_unit) {
    // Normalize inputs
    std::string from = normalize_unit(from_unit);
    std::string to   = normalize_unit(to_unit);

    // Look up unit types (throws on unknown unit)
    UnitType from_type = get_unit_type(from);
    UnitType to_type   = get_unit_type(to);

    // Check compatibility
    if (from_type != to_type) {
        throw std::invalid_argument(
            "Cannot convert between " + unit_type_to_string(from_type) +
            " and " + unit_type_to_string(to_type));
    }

    // Temperature: non-linear, handled separately
    if (from_type == UnitType::Temperature) {
        return convert_temperature(value, from, to);
    }

    // Count: dimensionless, pass through
    if (from_type == UnitType::Count) {
        return value;
    }

    // Volume / Weight: from_unit → base → to_unit
    double from_factor = UNIT_TABLE.at(from).to_base_factor;
    double to_factor   = UNIT_TABLE.at(to).to_base_factor;
    double base_value  = value * from_factor;
    return base_value / to_factor;
}

} // namespace lethimcook
